File: windows/src/MTGSheet.App/DeckView.cs
using Microsoft.UI.Xaml;
using Microsoft.UI.Xaml.Controls;
using Microsoft.UI.Xaml.Media;
using MTGSheet.Core;

namespace MTGSheet.App;

/// Phase 1: paste a Moxfield list, pick an art variant for every copy, download. Mirrors DeckView.swift.
public sealed class DeckView : UserControl
{
    public sealed class Face
    {
        public MPCFill.Query Query;
        public List<string> Results = [];
        public MPCFill.Card? Selected;
    }

    public sealed class Row
    {
        public required string Name;
        public int Copy = 1;
        public int Copies = 1;
        public bool Included = true;
        public required Face Front;
        public Face? Back;
        public string Title => Copies > 1 ? $"{Name} {Copy}/{Copies}" : Name;
    }

    private readonly Window window;
    private readonly Action<List<PrintCard>> onReady;
    private readonly TextBox list = new()
    {
        AcceptsReturn = true,
        TextWrapping = TextWrapping.NoWrap,
        FontFamily = new FontFamily("Consolas"),
        PlaceholderText = "1 Abrade\n11 Island",
    };
    private readonly Ui.WrapPanel grid = new();
    private readonly TextBlock status = Ui.Text("", dim: true);
    private readonly Slider tileSize = new() { Minimum = 90, Maximum = 300, Width = 120 };
    private readonly TextBlock counts = Ui.Text("");
    private readonly Button searchButton;
    private readonly Button downloadButton;

    private readonly List<Row> rows = [];
    private readonly Dictionary<string, MPCFill.Card> cardCache = [];
    private int[]? sources;
    private Dictionary<string, string>? dfcPairs;
    private bool busy;

    public DeckView(Window window, Action<List<PrintCard>> onReady)
    {
        this.window = window;
        this.onReady = onReady;

        searchButton = Ui.Button(Loc.Tr("Search MPCFill"), async (_, _) => await SearchAsync());
        downloadButton = Ui.Button(Loc.Tr("Download and lay out"), async (_, _) => await DownloadAsync());
        tileSize.Value = AppSettings.Current.TileSize;
        tileSize.ValueChanged += (_, e) =>
        {
            AppSettings.Current.TileSize = e.NewValue;
            AppSettings.Current.Save();
            grid.ItemWidth = e.NewValue;
            Rebuild();
        };
        grid.ItemWidth = tileSize.Value;

        var left = new Grid { Padding = new Thickness(10), Width = 320 };
        left.RowDefinitions.Add(new RowDefinition { Height = GridLength.Auto });
        left.RowDefinitions.Add(new RowDefinition { Height = new GridLength(1, GridUnitType.Star) });
        left.RowDefinitions.Add(new RowDefinition { Height = GridLength.Auto });
        var title = Ui.Text(Loc.Tr("Deck list"), 16, bold: true);
        Grid.SetRow(title, 0);
        Grid.SetRow(list, 1);
        Grid.SetRow(searchButton, 2);
        left.Children.Add(title);
        left.Children.Add(list);
        left.Children.Add(searchButton);

        var bottom = new Grid { Padding = new Thickness(10) };
        bottom.ColumnDefinitions.Add(new ColumnDefinition { Width = new GridLength(1, GridUnitType.Star) });
        bottom.ColumnDefinitions.Add(new ColumnDefinition { Width = GridLength.Auto });
        Grid.SetColumn(status, 0);
        var right = Ui.Row(tileSize, counts, downloadButton);
        Grid.SetColumn(right, 1);
        bottom.Children.Add(status);
        bottom.Children.Add(right);

        var rightSide = new Grid();
        rightSide.RowDefinitions.Add(new RowDefinition { Height = new GridLength(1, GridUnitType.Star) });
        rightSide.RowDefinitions.Add(new RowDefinition { Height = GridLength.Auto });
        var scroller = new ScrollViewer { Content = grid, Padding = new Thickness(12) };
        Grid.SetRow(scroller, 0);
        Grid.SetRow(bottom, 1);
        rightSide.Children.Add(scroller);
        rightSide.Children.Add(bottom);

        var root = new Grid();
        root.ColumnDefinitions.Add(new ColumnDefinition { Width = GridLength.Auto });
        root.ColumnDefinitions.Add(new ColumnDefinition { Width = new GridLength(1, GridUnitType.Star) });
        Grid.SetColumn(left, 0);
        Grid.SetColumn(rightSide, 1);
        root.Children.Add(left);
        root.Children.Add(rightSide);
        Content = root;

        UpdateCounts();
    }

    private void SetBusy(bool value)
    {
        busy = value;
        searchButton.IsEnabled = !value;
        downloadButton.IsEnabled = !value && rows.Any(r => r.Included && r.Front.Selected is not null);
    }

    private async Task SearchAsync()
    {
        if (busy) return;
        SetBusy(true);
        status.Text = Loc.Tr("Searching MPCFill…");
        try
        {
            sources ??= await MPCFill.SourceIdsAsync();
            dfcPairs ??= await MPCFill.DfcPairsAsync();
            var entries = MPCFill.ParseDecklist(list.Text, dfcPairs);
            var queries = entries.SelectMany(e => e.Back is { } back ? new[] { e.Front, back } : [e.Front]).ToList();
            var hits = await MPCFill.SearchAsync(queries, sources);

            rows.Clear();
            foreach (var entry in entries)
                for (var copy = 1; copy <= entry.Quantity; copy++)
                    rows.Add(new Row
                    {
                        Name = entry.Name,
                        Copy = copy,
                        Copies = entry.Quantity,
                        Front = new Face { Query = entry.Front, Results = hits.GetValueOrDefault(entry.Front, []) },
                        Back = entry.Back is { } back
                            ? new Face { Query = back, Results = hits.GetValueOrDefault(back, []) }
                            : null,
                    });

            // Best variant of every face preselected, like mpcfill.com.
            var firsts = rows.SelectMany(r => new[] { r.Front.Results.FirstOrDefault(), r.Back?.Results.FirstOrDefault() })
                             .Where(id => id is not null).Select(id => id!).Distinct().ToList();
            var details = await CardsAsync(firsts);
            foreach (var row in rows)
            {
                row.Front.Selected = Pick(row.Front, details);
                if (row.Back is not null) row.Back.Selected = Pick(row.Back, details);
                row.Included = row.Front.Selected is not null;
            }

            var missing = rows.Where(r => r.Front.Selected is null && r.Copy == 1).Select(r => r.Name).ToList();
            status.Text = missing.Count == 0
                ? Loc.Tr("Cards found: %d.", entries.Count)
                : Loc.Tr("Not found: %@", string.Join(", ", missing));
            Rebuild();
        }
        catch (Exception error)
        {
            status.Text = Loc.Tr("Error: %@", error.Message);
        }
        finally
        {
            SetBusy(false);
        }
    }

    private static MPCFill.Card? Pick(Face face, IReadOnlyDictionary<string, MPCFill.Card> details) =>
        face.Results.FirstOrDefault() is { } id && details.TryGetValue(id, out var card) ? card : null;

    private async Task<Dictionary<string, MPCFill.Card>> CardsAsync(IReadOnlyList<string> identifiers)
    {
        var missing = identifiers.Where(id => !cardCache.ContainsKey(id)).ToList();
        if (missing.Count > 0)
            foreach (var (id, card) in await MPCFill.CardsAsync(missing))
                cardCache[id] = card;
        return cardCache;
    }

    private void Rebuild()
    {
        grid.Children.Clear();
        foreach (var row in rows)
        {
            grid.Children.Add(Tile(row, back: false));
            if (row.Back is not null) grid.Children.Add(Tile(row, back: true));
        }
        UpdateCounts();
    }

    private UIElement Tile(Row row, bool back)
    {
        var face = back ? row.Back! : row.Front;
        var width = tileSize.Value;
        var image = new Image
        {
            Source = Ui.FromUrl(face.Selected?.SmallThumbnailUrl),
            Width = width,
            Height = width * 1122 / 822,
            Stretch = Stretch.Uniform,
            Opacity = row.Included ? 1 : 0.35,
        };
        var button = new Button { Content = image, Padding = new Thickness(0), IsEnabled = face.Results.Count > 0 };
        ToolTipService.SetToolTip(button, Loc.Tr("Choose the variant"));
        button.Click += (_, _) => new CardPickerWindow(
            back ? Loc.Tr("Back: %@", face.Selected?.Name ?? "") : row.Title,
            face.Selected?.Identifier,
            () => VariantsAsync(face),
            card => { face.Selected = card; Rebuild(); }).Activate();

        var tile = Ui.Column(button, Ui.Text(back ? Loc.Tr("Back: %@", face.Selected?.Name ?? "") : row.Title, 12, bold: true));
        tile.Children.Add(face.Selected is { } card
            ? Ui.Text(Loc.Tr("%@ · %d DPI · %d var.", card.SourceName, card.Dpi, face.Results.Count), 11, dim: true)
            : Ui.Text(Loc.Tr("Not found"), 11).With(t => t.Foreground = new SolidColorBrush(Microsoft.UI.Colors.IndianRed)));
        if (!back)
        {
            var include = new CheckBox { IsChecked = row.Included, IsEnabled = row.Front.Selected is not null };
            ToolTipService.SetToolTip(include, Loc.Tr("Print this card"));
            include.Checked += (_, _) => { row.Included = true; UpdateCounts(); };
            include.Unchecked += (_, _) => { row.Included = false; UpdateCounts(); };
            tile.Children.Add(include);
        }
        tile.Width = width;
        return tile;
    }

    private async Task<List<MPCFill.Card>> VariantsAsync(Face face)
    {
        var details = await CardsAsync(face.Results);
        return face.Results.Where(details.ContainsKey).Select(id => details[id]).ToList();
    }

    private void UpdateCounts()
    {
        var chosen = rows.Where(r => r.Included && r.Front.Selected is not null).ToList();
        counts.Text = Loc.Tr("Cards: %d · Double-sided: %d",
            chosen.Count(r => r.Back is null), chosen.Count(r => r.Back is not null));
        downloadButton.IsEnabled = !busy && chosen.Count > 0;
    }

    /// Downloads the chosen variants, four at a time, and hands one print card per copy to phase 2.
    private async Task DownloadAsync()
    {
        if (busy) return;
        SetBusy(true);
        var chosen = rows.Where(r => r.Included && r.Front.Selected is not null).ToList();
        var unique = chosen.SelectMany(r => new[] { r.Front.Selected, r.Back?.Selected })
                           .Where(c => c is not null).Select(c => c!).Distinct().ToList();
        var files = new Dictionary<string, string>();
        var done = 0;
        try
        {
            using var limit = new SemaphoreSlim(4);
            await Task.WhenAll(unique.Select(async card =>
            {
                await limit.WaitAsync();
                try
                {
                    var file = await MPCFill.DownloadAsync(card, Paths.Cache);
                    lock (files)
                    {
                        files[card.Identifier] = file;
                        done++;
                    }
                    status.Text = Loc.Tr("Downloading %d/%d…", done, unique.Count);
                }
                finally { limit.Release(); }
            }));

            var cards = chosen
                .Where(r => r.Front.Selected is not null && files.ContainsKey(r.Front.Selected!.Identifier))
                .Select(r => new PrintCard(files[r.Front.Selected!.Identifier],
                    r.Back?.Selected is { } back && files.TryGetValue(back.Identifier, out var path) ? path : null))
                .ToList();
            status.Text = Loc.Tr("Downloaded %d images.", unique.Count);
            onReady(cards);
        }
        catch (Exception error)
        {
            status.Text = Loc.Tr("Error: %@", error.Message);
        }
        finally
        {
            SetBusy(false);
        }
    }
}

File: Sources/MTGSheetOptimizer/DeckView.swift
import SwiftUI

/// Images picked in phase 1, handed to the layout phase.
struct DeckFiles {
    var cards: [URL]         // one entry per copy, in deck order
    var doubleSided: [URL]   // both faces of double-faced cards, exported as singles
}

let imageCacheDir = FileManager.default.urls(for: .cachesDirectory, in: .userDomainMask)[0]
    .appendingPathComponent("MTG Sheet Optimizer", isDirectory: true)

// MARK: - Model

@MainActor @Observable
final class DeckModel {
    struct Face {
        var query: MPCFill.Query
        var results: [String] = []   // variant identifiers, best first
        var selected: MPCFill.Card?
    }

    /// One physical card: "10 Island" becomes ten rows, each with its own variant.
    struct Row: Identifiable {
        let id = UUID()
        var name: String
        var copy = 1
        var copies = 1
        var included = true
        var front: Face
        var back: Face?

        var title: String { copies > 1 ? "\(name) \(copy)/\(copies)" : name }
    }

    var text = UserDefaults.standard.string(forKey: "deckText") ?? "" {
        didSet { UserDefaults.standard.set(text, forKey: "deckText") }
    }
    var rows: [Row] = []
    var status = ""
    var busy = false
    private var cardCache: [String: MPCFill.Card] = [:]
    private var sources: [Int]?
    private var dfcPairs: [String: String]?

    private var chosen: [Row] { rows.filter { $0.included && $0.front.selected != nil } }
    var pageCards: Int { chosen.filter { $0.back == nil }.count }
    var doubleSidedCards: Int { chosen.filter { $0.back != nil }.count }

    func search() async {
        busy = true
        defer { busy = false }
        status = tr("Searching MPCFill…")
        do {
            if sources == nil { sources = try await MPCFill.sourceIDs() }
            if dfcPairs == nil { dfcPairs = try await MPCFill.dfcPairs() }
            let entries = MPCFill.parseDecklist(text, dfcPairs: dfcPairs ?? [:])
            let hits = try await MPCFill.search(entries.flatMap { [$0.front] + ($0.back.map { [$0] } ?? []) },
                                                sources: sources ?? [])
            var newRows = entries.flatMap { e in
                (1...e.quantity).map { n in
                    Row(name: e.name, copy: n, copies: e.quantity,
                        front: Face(query: e.front, results: hits[e.front] ?? []),
                        back: e.back.map { Face(query: $0, results: hits[$0] ?? []) })
                }
            }
            // Best variant of every face preselected, like mpcfill.com.
            let firsts = newRows.flatMap { [$0.front.results.first, $0.back?.results.first] }.compactMap { $0 }
            let details = try await cards(Array(Set(firsts)))
            for i in newRows.indices {
                let front = newRows[i].front.results.first.flatMap { details[$0] }
                let back = newRows[i].back?.results.first.flatMap { details[$0] }
                newRows[i].front.selected = front
                newRows[i].back?.selected = back
                newRows[i].included = front != nil
            }
            rows = newRows
            let missing = rows.filter { $0.front.selected == nil && $0.copy == 1 }.map(\.name)
            status = missing.isEmpty ? tr("Found all %d cards.", entries.count)
                                     : tr("Not found: %@", missing.joined(separator: ", "))
        } catch {
            status = tr("Error: %@", error.localizedDescription)
        }
    }

    /// Details of `ids`, in order, each fetched once.
    func variants(_ ids: [String]) async throws -> [MPCFill.Card] {
        let details = try await cards(ids)
        return ids.compactMap { details[$0] }
    }

    private func cards(_ ids: [String]) async throws -> [String: MPCFill.Card] {
        let missing = ids.filter { cardCache[$0] == nil }
        if !missing.isEmpty { cardCache.merge(try await MPCFill.cards(missing)) { a, _ in a } }
        return cardCache
    }

    /// Downloads the chosen variants (4 at a time) and returns one entry per card.
    func download() async -> DeckFiles? {
        busy = true
        defer { busy = false }
        let rows = chosen
        let unique = Array(Set(rows.flatMap { [$0.front.selected, $0.back?.selected] }.compactMap { $0 }))
        var files: [String: URL] = [:]
        do {
            try await withThrowingTaskGroup(of: (String, URL).self) { group in
                for (i, card) in unique.enumerated() {
                    if i >= 4, let done = try await group.next() {
                        files[done.0] = done.1
                        status = tr("Downloading %d/%d…", files.count, unique.count)
                    }
                    group.addTask { (card.identifier, try await MPCFill.download(card, to: imageCacheDir)) }
                }
                for try await done in group {
                    files[done.0] = done.1
                    status = tr("Downloading %d/%d…", files.count, unique.count)
                }
            }
        } catch {
            status = tr("Error: %@", error.localizedDescription)
            return nil
        }

        var out = DeckFiles(cards: [], doubleSided: [])
        for row in rows {
            let faces = [row.front.selected, row.back?.selected].compactMap { $0.flatMap { files[$0.identifier] } }
            if row.back == nil { out.cards += faces } else { out.doubleSided += faces }
        }
        status = tr("Downloaded %d images.", unique.count)
        return out
    }
}

// MARK: - Views

struct DeckView: View {
    @Bindable var deck: DeckModel
    let onReady: (DeckFiles) -> Void
    @State private var picking: Pick?
    @AppStorage("tileSize") private var tileSize = 150.0

    struct Pick: Identifiable {
        let row: UUID
        let back: Bool
        var id: String { "\(row)-\(back)" }
    }

    var body: some View {
        HSplitView {
            VStack(alignment: .leading, spacing: 8) {
                Text(tr("Deck list")).font(.headline)
                TextEditor(text: $deck.text)
                    .font(.system(.body, design: .monospaced))
                    .border(Color.secondary.opacity(0.3))
                Button(tr("Search MPCFill")) { Task { await deck.search() } }
                    .disabled(deck.busy || deck.text.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty)
            }
            .padding(10)
            .frame(minWidth: 220, idealWidth: 280, maxWidth: 420)

            VStack(spacing: 0) {
                if deck.rows.isEmpty {
                    Text(tr("Paste the list and press \"Search MPCFill\"."))
                        .foregroundStyle(.secondary)
                        .frame(maxWidth: .infinity, maxHeight: .infinity)
                } else {
                    ScrollView {
                        LazyVGrid(columns: [GridItem(.adaptive(minimum: tileSize), spacing: 12)], spacing: 16) {
                            ForEach($deck.rows) { $row in
                                FaceTile(row: $row, back: false) { picking = Pick(row: row.id, back: false) }
                                if row.back != nil {
                                    FaceTile(row: $row, back: true) { picking = Pick(row: row.id, back: true) }
                                }
                            }
                        }
                        .padding(12)
                    }
                }
                Divider()
                HStack {
                    if deck.busy { ProgressView().controlSize(.small) }
                    Text(deck.status).foregroundStyle(.secondary).lineLimit(1)
                    Spacer()
                    Image(systemName: "square.grid.3x3").foregroundStyle(.secondary)
                    Slider(value: $tileSize, in: 90...300)
                        .frame(width: 120)
                        .help(tr("Card size"))
                    Image(systemName: "square.grid.2x2").foregroundStyle(.secondary)
                    Text(tr("%d cards, %d double-sided", deck.pageCards, deck.doubleSidedCards))
                    Button(tr("Download and lay out")) {
                        Task { if let files = await deck.download() { onReady(files) } }
                    }
                    .disabled(deck.busy || deck.pageCards + deck.doubleSidedCards == 0)
                }
                .padding(10)
            }
            .frame(minWidth: 500)
        }
        .sheet(item: $picking) { pick in
            if let i = deck.rows.firstIndex(where: { $0.id == pick.row }),
               let face = pick.back ? deck.rows[i].back : deck.rows[i].front as DeckModel.Face? {
                VariantPicker(title: deck.rows[i].title, face: face, deck: deck) { card in
                    if pick.back { deck.rows[i].back?.selected = card } else { deck.rows[i].front.selected = card }
                }
            }
        }
    }
}

private struct FaceTile: View {
    @Binding var row: DeckModel.Row
    let back: Bool
    let pick: () -> Void

    var body: some View {
        let face = back ? row.back : row.front
        VStack(spacing: 4) {
            Button(action: pick) { Thumbnail(url: face?.selected?.smallThumbnailUrl) }
                .buttonStyle(.plain)
                .disabled(face?.results.isEmpty ?? true)
                .opacity(row.included ? 1 : 0.35)
                .help(tr("Choose the variant"))
            Text(back ? tr("Back: %@", face?.selected?.name ?? "") : row.title)
                .font(.caption.bold())
                .lineLimit(1)
            if let face, let card = face.selected {
                Text(tr("%@ · %d DPI · %d variants", card.sourceName, card.dpi, face.results.count))
                    .font(.caption2)
                    .foregroundStyle(.secondary)
                    .lineLimit(1)
            } else {
                Text(tr("Not found")).font(.caption2).foregroundStyle(.red)
            }
            if !back {
                Toggle("", isOn: $row.included)
                    .labelsHidden()
                    .disabled(row.front.selected == nil)
                    .help(tr("Print this card"))
            }
        }
    }
}

private struct Thumbnail: View {
    let url: String?

    var body: some View {
        AsyncImage(url: url.flatMap(URL.init(string:))) { image in
            image.resizable().scaledToFit()
        } placeholder: {
            RoundedRectangle(cornerRadius: 6)
                .fill(.quaternary)
                .overlay { if url != nil { ProgressView().controlSize(.small) } }
        }
        .aspectRatio(822.0 / 1122.0, contentMode: .fit)
        .clipShape(RoundedRectangle(cornerRadius: 6))
    }
}

private struct VariantPicker: View {
    let title: String
    let face: DeckModel.Face
    let deck: DeckModel
    let onPick: (MPCFill.Card) -> Void
    @Environment(\.dismiss) private var dismiss
    @AppStorage("tileSize") private var tileSize = 150.0
    @State private var cards: [MPCFill.Card] = []
    @State private var error: String?

    var body: some View {
        VStack(spacing: 0) {
            HStack {
                Text(title).font(.headline)
                Text(tr("%d variants", face.results.count)).foregroundStyle(.secondary)
                Spacer()
                Button(tr("Close")) { dismiss() }.keyboardShortcut(.cancelAction)
            }
            .padding(12)
            Divider()
            if let error {
                Text(error)
                    .foregroundStyle(.red)
                    .frame(maxWidth: .infinity, maxHeight: .infinity)
            } else if cards.isEmpty {
                ProgressView(tr("Loading %d variants…", face.results.count))
                    .frame(maxWidth: .infinity, maxHeight: .infinity)
            } else {
                ScrollView {
                    LazyVGrid(columns: [GridItem(.adaptive(minimum: tileSize), spacing: 12)], spacing: 14) {
                        ForEach(cards) { card in
                            Button {
                                onPick(card)
                                dismiss()
                            } label: {
                                VStack(spacing: 3) {
                                    Thumbnail(url: card.smallThumbnailUrl)
                                        .overlay {
                                            if card.identifier == face.selected?.identifier {
                                                RoundedRectangle(cornerRadius: 6).stroke(Color.accentColor, lineWidth: 3)
                                            }
                                        }
                                    Text(card.sourceName).font(.caption).lineLimit(1)
                                    Text("\(card.dpi) DPI · \(String(format: "%.1f", Double(card.size) / 1_000_000)) MB")
                                        .font(.caption2)
                                        .foregroundStyle(.secondary)
                                }
                            }
                            .buttonStyle(.plain)
                        }
                    }
                    .padding(12)
                }
            }
        }
        .frame(minWidth: 600, idealWidth: 820, minHeight: 440, idealHeight: 640)
        .task {
            do { cards = try await deck.variants(face.results) } catch { self.error = error.localizedDescription }
        }
    }
}
